static/panes/device-view.ts

```diff
@@ -0,0 +1,378 @@
+        this.selectize.on('change', this.onDeviceSelect.bind(this));
+        this.paneRenaming.on('renamePane', this.updateState.bind(this));
+
+        this.container.on('destroy', this.close, this);
+
+        this.eventHub.on('compileResult', this.onCompileResult, this);
+        this.eventHub.on('compiler', this.onCompiler, this);
+        this.eventHub.on('colours', this.onColours, this);
+        this.eventHub.on('panesLinkLine', this.onPanesLinkLine, this);
+        this.eventHub.on('compilerClose', this.onCompilerClose, this);
+        this.eventHub.on('settingsChange', this.onSettingsChange, this);
+        this.eventHub.emit('deviceViewOpened', this.compilerInfo.compilerId);
+        this.eventHub.emit('requestSettings');
+
+        this.container.on('resize', this.resize, this);
+        this.container.on('shown', this.resize, this);
+    }
+
+    override onCompileResult(id: number, compiler: any, result: any): void {
+        if (this.compilerInfo.compilerId !== id) return;
+        this.devices = result.devices;
+        let deviceNames: string[] = [];
+        if (!this.devices) {
+            this.showDeviceAsmResults([{text: '<No output>'}]);
+        } else {
+            deviceNames = Object.keys(this.devices);
+        }
+
+        this.makeDeviceSelector(deviceNames);
+        this.updateDeviceAsm();
+
+        // Why call this explicitly instead of just listening to the "colours" event?
+        // Because the recolouring happens before this editors value is set using "showDeviceAsmResults".
+        this.onColours(this.compilerInfo.compilerId, this.lastColours, this.lastColourScheme);
+    }
+
+    makeDeviceSelector(deviceNames: string[]): void {
+        const selectize = this.selectize;
+
+        for (const key in selectize.options) {
+            if (deviceNames.indexOf(selectize.options[key].name) === -1) {
+                selectize.removeOption(selectize.options[key].name);
+            }
+        }
+
+        deviceNames.forEach(deviceName => {
+            selectize.addOption({name: deviceName});
+        });
+
+        if (!this.selectedDevice && deviceNames.length > 0) {
+            this.selectedDevice = deviceNames[0];
+            selectize.setValue(this.selectedDevice, true);
+        } else if (this.selectedDevice && deviceNames.indexOf(this.selectedDevice) === -1) {
+            selectize.clear(true);
+            this.showDeviceAsmResults([{text: '<Device ' + this.selectedDevice + ' not found>'}]);
+        } else {
+            selectize.setValue(this.selectedDevice, true);
+            this.updateDeviceAsm();
+        }
+    }
+
+    onDeviceSelect(): void {
+        this.selectedDevice = this.selectize.getValue() as string;
+        this.updateState();
+        this.updateDeviceAsm();
+    }
+
+    updateDeviceAsm(): void {
+        // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+        if (this.selectedDevice && this.devices != null && this.devices[this.selectedDevice]) {
```

Comment:
To check that a property is in a Record (JS object), it's better to use `this.selectedDevice in this.devices`